tree/dataframe/inc/ROOT/RDF/RMergeableValue.hxx

```diff
@@ -0,0 +1,458 @@
+
+// Merge the values and get another mergeable back
+auto mergedptr = MergeValues(std::move(mergeableh1), std::move(mergeableh2));
+
+// Retrieve the merged TH1D object
+const auto &mergedhisto = mergedptr->GetValue();
+~~~
+
+Though this snippet can run on a single thread of a single machine, it is
+straightforward to generalize it to a distributed case, e.g. where `mergeableh1`
+and `mergeableh2` are created on separate machines and sent to a `reduce`
+process where the `MergeValues` function is called. The final user would then
+just be given the final merged result coming from `mergedptr->GetValue`.
+
+RMergeableValue is the base class for all the different specializations that may
+be needed according to the peculiarities of the result types. The following
+subclasses are currently available:
+
+| RMergeableValue Specialization | Corresponding RDataFrame Action(s)         |
+|--------------------------------|--------------------------------------------|
+| RMergeableFill                 | Graph, Histo{1,2,3}D, Profile{1,2}D, Stats |
+| RMergeableGenericAction        | Count, Max, Min, Sum                       |
+| RMergeableMean                 | Mean                                       |
+| RMergeableStdDev               | StdDev                                     |
+*/
+template <typename T>
+class RMergeableValue : public RMergeableValueBase {
+   // Friend function declaration
+   template <typename T1, typename... Ts>
+   friend std::unique_ptr<RMergeableValue<T1>> MergeValues(std::unique_ptr<RMergeableValue<T1>> OutputMergeable,
+                                                           std::unique_ptr<RMergeableValue<Ts>>... InputMergeables);
+   template <typename T1, typename... Ts>
+   friend void MergeValues(RMergeableValue<T1> &OutputMergeable, const RMergeableValue<Ts> &... InputMergeables);
+
+   /////////////////////////////////////////////////////////////////////////////
+   /// \brief Aggregate the information contained in another RMergeableValue
+   ///        into this.
+   ///
+   /// Virtual function reimplemented in all the subclasses.
+   ///
+   /// \note All the `Merge` methods in the RMergeableValue family are private.
+   /// To merge multiple RMergeableValue objects please use [MergeValues]
+   /// (namespaceROOT_1_1Detail_1_1RDF.html#af16fefbe2d120983123ddf8a1e137277).
+   virtual void Merge(const RMergeableValue<T> &) = 0;
+
+protected:
+   T fValue; ///< Result of the RDataFrame action wrapped by this mergeable.
+
+public:
+   /**
+      Constructor taking the action result by const reference. This involves a
+      copy of the result into the data member, but gives full ownership of data
+      to the mergeable.
+   */
+   RMergeableValue(const T &value) : fValue{value} {}
+
+   /**
+      Retrieve the value contained in this mergeable object. Meant to be called
+      after merging two ore more objects together to retrieve the final result.
+   */
+   const T &GetValue() const { return fValue; }
```

Comment:
so iiuc this method is only meant to be called on RMergeableValues that contain the result of a merge. That's easily checked with a boolean, and it might be an easy way to prevent misuse, what do you think?

Reply:
Fair enough yes, can  add a boolean to check `MergeValues` has been called at least once.
This wouldn't still account for all cases, e.g. a mergeable that needs to pass 2 or more rounds of merging before actually getting all the info (especially true for the `MergeValues` signature with the `unique_ptr`s). But I see the need for checking at least one merging has happened

Reply:
uhm ok up to you then, i don't know whether a check that might have false negatives is better than no check at all in this case

Reply:
Added fIsMerged data member to check that the mergeable has been merged at least once